src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp

```diff
@@ -49,14 +53,67 @@ double GoalieTactic::calculateRobotCost(const Robot &robot, const World &world)
     }
     else
     {
-        return 1000000;
+        return std::numeric_limits<int>::max() - 10;
+    }
+}
+
+std::optional<Point> GoalieTactic::restrainGoalieInRectangle(
+    Point goalie_desired_position, Rectangle goalie_restricted_area)
+{
+    Point clamped_goalie_pos;
+
+    // get the intersections to all 3 crease lines in question
+    auto width_x_goal    = lineIntersection(goalie_desired_position, field.friendlyGoal(),
+                                         goalie_restricted_area.neCorner(),
+                                         goalie_restricted_area.seCorner());
+    auto pos_side_x_goal = lineIntersection(goalie_desired_position, field.friendlyGoal(),
+                                            goalie_restricted_area.neCorner(),
+                                            goalie_restricted_area.nwCorner());
+    auto neg_side_x_goal = lineIntersection(goalie_desired_position, field.friendlyGoal(),
+                                            goalie_restricted_area.seCorner(),
+                                            goalie_restricted_area.swCorner());
+
+    // if the goalie restricted area already contains the point, then we are
+    // safe to move there.
+    if (goalie_restricted_area.containsPoint(goalie_desired_position))
+    {
+        return std::make_optional<Point>(goalie_desired_position);
+    }
+    // if the goalie position to the block cone position intersects the
+    // longest crease line which runs across the y axis, then intersect
+    else if (width_x_goal && width_x_goal->y() <= goalie_restricted_area.neCorner().y() &&
+             width_x_goal->y() >= goalie_restricted_area.seCorner().y())
+    {
+        return std::make_optional<Point>(*width_x_goal);
+    }
+
+    // if either two sides of the goal are intercepted, then use those positions
+    else if (pos_side_x_goal &&
+             pos_side_x_goal->x() <= goalie_restricted_area.neCorner().x() &&
+             pos_side_x_goal->x() >= goalie_restricted_area.nwCorner().x())
+    {
+        return std::make_optional<Point>(*pos_side_x_goal);
+    }
+    else if (neg_side_x_goal &&
+             neg_side_x_goal->x() <= goalie_restricted_area.seCorner().x() &&
+             neg_side_x_goal->x() >= goalie_restricted_area.swCorner().x())
+    {
+        return std::make_optional<Point>(*neg_side_x_goal);
+    }
+
+    // if there are no intersections, then we are out of luck
+    else
+    {
+        return std::nullopt;
```

Comment:
this can happen if the ball is behind the net, then we don't have an intersection with the goalie rectangle, in this case it defaults to the closest post. 

So this case will almost always be reached when the ball is at a super sharp angle or is behind the net and has no direct shot on the net.

Reply:
@garethellis0 this is expected to happen, should I put a `LOG(INFO)` here? It happens every-time there is no intersection with the ball and the net.

Reply:
Oh sorry, I left a comment here, not sure why it didn't get posted. I think just expanding the above comment to something like:
```
// if there are no intersections (ex. ball behind net), then we are out of luck. 
```

Reply:
fixed